apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py

```diff
@@ -770,6 +770,11 @@ def _perform_gw_port_operations(self, context, port):
         network_tenant = self._get_network_aci_tenant(network)
         router = self.l3_plugin.get_router(context._plugin_context, router_id)
 
+        # Other L3 plugins (e.g. ASR) create db-only routers, which
+        # are indicated by an empty string tenant. Don't do anything
+        # for these devices
+        if router['tenant_id'] == '':
+            return
         vrf = self._get_tenant_vrf(router['tenant_id'])
```

Comment:
i think you need similar checks in _check_gw_port_operation() and _delete_contract()